visualization/panda/panda3d_utils.py

```diff
@@ -0,0 +1,239 @@
+        # setup render 2d
+        self.render2d = NodePath("extra_win_render2d")
+        self.render2d.setDepthTest(0)
+        self.render2d.setDepthWrite(0)
+
+        self.win = base.openWindow(props=WindowProperties(base.win.getProperties()),
+                                   makeCamera=False,
+                                   scene=self.render,
+                                   requireWindow=True, )
+
+        # set window background to white
+        base.setBackgroundColor(r=1, g=1, b=1, win=self.win)
+        # set window title and window's dimension
+        self.set_win_props(title=window_title,
+                           size=(w, h), )
+        # set len for the camera and set the camera for the new window
+        lens = PerspectiveLens()
+        lens.setFov(fov)
+        lens.setNearFar(0.001, 5000.0)
+        if lens_type == "orthographic":
+            lens = OrthographicLens()
+            lens.setFilmSize(1, 1)
+        # make aspect ratio looks same as base window
+        aspect_ratio = base.getAspectRatio()
+        lens.setAspectRatio(aspect_ratio)
+        self.cam = base.makeCamera(self.win, scene=self.render, )  # can also be found in base.camList
+        self.cam.reparentTo(self.render)
+        self.cam.setPos(Point3(cam_pos[0], cam_pos[1], cam_pos[2]))
+        self.cam.lookAt(Point3(lookat_pos[0], lookat_pos[1], lookat_pos[2]), Vec3(up[0], up[1], up[2]))
+        self.cam.node().setLens(lens)  # use same len as sys
+        # set up cartoon effect
+        self._separation = 1
+        self.filter = flt.Filter(self.win, self.cam)
+        self.filter.setCartoonInk(separation=self._separation)
+
+        # camera in camera 2d
+        self.cam2d = base.makeCamera2d(self.win, )
+        self.cam2d.reparentTo(self.render2d)
+        # attach GPTop to the render2d to make sure the DirectGui can be used
+        self.aspect2d = self.render2d.attachNewNode(PGTop("aspect2d"))
+        # self.aspect2d.setScale(1.0 / aspect_ratio, 1.0, 1.0)
+
+        # setup mouse for the new window
+        # name of mouse watcher is to adapt to the name in the input manager
+        self.mouse_thrower = base.setupMouse(self.win, fMultiWin=True)
+        self.mouseWatcher = self.mouse_thrower.getParent()
+        self.mouseWatcherNode = self.mouseWatcher.node()
+        self.aspect2d.node().setMouseWatcher(self.mouseWatcherNode)
+        # self.mouseWatcherNode.addRegion(PGMouseWatcherBackground())
+
+        # setup input manager
+        self.inputmgr = im.InputManager(self, lookatpos=lookat_pos)
+
+        # copy attributes and functions from base
+        ## change the bound function to a function, and bind to `self` to become a unbound function
+        self._interaction_update = functools.partial(base._interaction_update.__func__, self)
+        self.p3dh = base.p3dh
+
+        base.taskMgr.add(self._interaction_update, "interaction_extra_window", appendTask=True)
+
+    def getAspectRatio(self):
```

Comment:
use the pythonic name style

Reply:
I see. Since `self._interaction_update` uses the function `getAspectRatio`. So I have to use the name `getAspectRatio` here to keep the mouse operation for the ExtraWindow works well. I can overwrite the `_interaction_update`, but I think that makes the program tedious.